// payjp/built/charge.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const resource_1 = require("./resource");
class Charges extends resource_1.default {
    resource;
    constructor(payjp) {
        super(payjp);
        this.resource = 'charges';
    }
    list(query = {}) {
        return this.request('GET', this.resource, query);
    }
    create(query) {
        return this.request('POST', this.resource, query);
    }
    retrieve(id) {
        return this.request('GET', `${this.resource}/${id}`);
    }
    update(id, query = {}) {
        return this.request('POST', `${this.resource}/${id}`, query);
    }
    refund(id, query = {}) {
        return this.request('POST', `${this.resource}/${id}/refund`, query);
    }
    reauth(id, query = {}) {
        return this.request('POST', `${this.resource}/${id}/reauth`, query);
    }
    capture(id, query = {}) {
        return this.request('POST', `${this.resource}/${id}/capture`, query);
    }
    tds_finish(id) {
        return this.request('POST', `${this.resource}/${id}/tds_finish`);
    }
}
exports.default = Charges;
