// payjp/built/customer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const resource_1 = require("./resource");
const card_1 = require("./card");
class CustomerSubscriptions extends resource_1.default {
    resource;
    constructor(payjp) {
        super(payjp);
        this.resource = 'subscriptions';
    }
    list(id, query = {}) {
        return this.request('GET', `customers/${id}/${this.resource}`, query);
    }
    retrieve(id, subscriptionId) {
        return this.request('GET', `customers/${id}/${this.resource}/${subscriptionId}`);
    }
}
class Customers extends resource_1.default {
    resource;
    cards;
    subscriptions;
    constructor(payjp) {
        super(payjp);
        this.resource = 'customers';
        this.cards = new card_1.default(payjp);
        this.subscriptions = new CustomerSubscriptions(payjp);
    }
    list(query = {}) {
        return this.request('GET', this.resource, query);
    }
    create(query = {}) {
        return this.request('POST', this.resource, query);
    }
    retrieve(id) {
        return this.request('GET', `${this.resource}/${id}`);
    }
    update(id, query = {}) {
        return this.request('POST', `${this.resource}/${id}`, query);
    }
    delete(id) {
        return this.request('DELETE', `${this.resource}/${id}`);
    }
}
exports.default = Customers;
